llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmTypeCheck.cpp

```diff
@@ -371,12 +371,22 @@ bool WebAssemblyAsmTypeCheck::checkTryTable(SMLoc ErrorLoc,
     if (Level < BlockInfoStack.size()) {
       const auto &DestBlockInfo =
           BlockInfoStack[BlockInfoStack.size() - Level - 1];
-      if (compareTypes(SentTypes, DestBlockInfo.Sig.Returns)) {
-        std::string ErrorMsg =
-            ErrorMsgBase + "type mismatch, catch tag type is " +
-            getTypesString(SentTypes) + ", but destination's type is " +
-            getTypesString(DestBlockInfo.Sig.Returns);
-        Error |= typeError(ErrorLoc, ErrorMsg);
+      if (DestBlockInfo.IsLoop) {
+        if (compareTypes(SentTypes, DestBlockInfo.Sig.Params)) {
+          std::string ErrorMsg =
+              ErrorMsgBase + "type mismatch, catch tag type is " +
+              getTypesString(SentTypes) + ", but destination's type is " +
+              getTypesString(DestBlockInfo.Sig.Params);
+          Error |= typeError(ErrorLoc, ErrorMsg);
+        }
+      } else {
+        if (compareTypes(SentTypes, DestBlockInfo.Sig.Returns)) {
+          std::string ErrorMsg =
+              ErrorMsgBase + "type mismatch, catch tag type is " +
+              getTypesString(SentTypes) + ", but destination's type is " +
+              getTypesString(DestBlockInfo.Sig.Returns);
+          Error |= typeError(ErrorLoc, ErrorMsg);
+        }
```

Comment:
This is kind of repetitive. I wanted something like
```cpp
ArrayRef DestTypes = DestBlockInfo.IsLoop ? DestBlockInfo.Sig.Params : DestBlockInfo.Sig.Returns;
if (compares(SentTypes, DestTypes)) {
  ...
}
```
But this was not possible because the types of `WasmSignature::Params` and `WasmSignature::Returns` are different:
https://github.com/llvm/llvm-project/blob/a98466ad8967f9ad9cd79157a7aed8ade5d65a7a/llvm/include/llvm/BinaryFormat/Wasm.h#L496-L497

Reply:
Hm... could you maybe upcast to `SmallVectorImpl` somewhere?

Reply:
Neither of
```cpp
ArrayRef<wasm::ValType> DestTypes = DestBlockInfo.IsLoop ? DestBlockInfo.Sig.Params : DestBlockInfo.Sig.Returns;
```
and
```cpp
SmallVectorImpl<wasm::ValType> DestTypes = DestBlockInfo.IsLoop ? DestBlockInfo.Sig.Params : DestBlockInfo.Sig.Returns;
```
works, saying things like
```console
.../llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmTypeCheck.cpp:375:32: error: incompatible operand types ('const SmallVector<[...], 4>' and 'const SmallVector<[...], 1>')
  375 |           DestBlockInfo.IsLoop ? DestBlockInfo.Sig.Params
      |                                ^ ~~~~~~~~~~~~~~~~~~~~~~~~
  376 |                                : DestBlockInfo.Sig.Returns;
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~
```

But this seems to work:
```cpp
      ArrayRef<wasm::ValType> DestTypes;
      if (DestBlockInfo.IsLoop)
        DestTypes = DestBlockInfo.Sig.Params;
      else
        DestTypes = DestBlockInfo.Sig.Returns;
      if (compareTypes(SentTypes, DestTypes)) {
        ...
```
Replacing `ArrayRef` with `SmallVectorImpl` here doesn't work because `SmallVectorImpl` doesn't seem to have its own default constructor.

Reply:
Got it. I wonder if you could upcast both sides of the `:` in the ternary to SmallVectorImpl... but that doesn't really seem any better than what you have here.